internal/cluster/cmd/clusterd/contract.go

```diff
@@ -0,0 +1,489 @@
+	}
+
+	renterPrivateKey := types.GeneratePrivateKey()
+	renterAddr := types.StandardUnlockHash(renterPrivateKey.PublicKey())
+	hostPrivateKey := types.GeneratePrivateKey()
+	hostAddr := types.StandardUnlockHash(hostPrivateKey.PublicKey())
+
+	// create a storage contract
+	contractPayout := types.Siacoins(10)
+	height := cm.Tip().Height
+	// this contract will expire
+	fc1 := types.V2FileContract{
+		RevisionNumber:   0,
+		Capacity:         4096,
+		Filesize:         0,
+		ProofHeight:      height + 1,
+		ExpirationHeight: height + 2,
+		RenterOutput:     types.SiacoinOutput{Address: renterAddr, Value: contractPayout},
+		HostOutput:       types.SiacoinOutput{Address: hostAddr, Value: contractPayout},
+		MissedHostValue:  types.ZeroCurrency,
+		TotalCollateral:  contractPayout,
+		RenterPublicKey:  renterPrivateKey.PublicKey(),
+		HostPublicKey:    hostPrivateKey.PublicKey(),
+	}
+	fcOut := fc1.RenterOutput.Value.Add(fc1.HostOutput.Value).Add(cm.TipState().V2FileContractTax(fc1))
+
+	// this contract will be revised and then resolved with storage proof
+	fc2 := fc1
+	fc2.ProofHeight += 1
+	fc2.ExpirationHeight += 10
+	fc2.RevisionNumber = 1
+
+	// this contract will be renewed
+	fc3 := fc1
+	fc3.ProofHeight += 10000
+	fc3.ExpirationHeight += 10000
+	fc3.RevisionNumber = 2
+
+	txn := types.V2Transaction{
+		FileContracts: []types.V2FileContract{fc1, fc2, fc3},
+	}
+	for i := range txn.FileContracts {
+		cs := cm.TipState()
+		txn.FileContracts[i].RenterSignature = renterPrivateKey.SignHash(cs.ContractSigHash(txn.FileContracts[i]))
+		txn.FileContracts[i].HostSignature = hostPrivateKey.SignHash(cs.ContractSigHash(txn.FileContracts[i]))
+	}
+	basis, toSign, err := w.FundV2Transaction(&txn, fcOut.Mul64(3), false)
+	if err != nil {
+		log.Panic("Failed to fund contract creation txn", zap.Error(err))
+	}
+	w.SignV2Inputs(&txn, toSign)
+
+	if _, err := cm.AddV2PoolTransactions(basis, []types.V2Transaction{txn}); err != nil {
+		log.Panic("Failed to add contract creation txn to pool", zap.Error(err))
+	}
+	tip := cm.Tip()
+	mineBlocks(log, cm, 1)
+
+	if err := syncWallet(cm, w, ws); err != nil {
+		log.Panic("Failed to scan wallet", zap.Error(err))
+	}
+
+	_, applied, err := cm.UpdatesSince(tip, 1000)
+	if err != nil {
+		log.Panic("Failed to retrieve chain updates", zap.Error(err))
+	}
+
+	// get the confirmed file contract element
+	var fces []types.V2FileContractElement
+	for _, fce := range applied[0].V2FileContractElementDiffs() {
+		fces = append(fces, fce.V2FileContractElement)
+	}
+	if len(fces) != 3 {
+		log.Panic("Wrong number of contracts in block, expected 3, got", zap.Int("len", len(fces)))
+	}
+	// sort by revision number so fc[0] = fc1, fc[1] = fc2, ...
+	sort.Slice(fces, func(i, j int) bool {
+		return fces[i].V2FileContract.RevisionNumber < fces[j].V2FileContract.RevisionNumber
+	})
+	for _, cau := range applied[1:] {
+		for i := range fces {
+			cau.UpdateElementProof(&fces[i].StateElement)
```

Comment:
Could you avoid this manual proof updating by asking the explorer for the latest element?